### ISSUE TITLE
Double free on Array

### ISSUE THREAD

**Issue by yotam5** [noib3/nvim-oxi#78]
Im trying to handle the complete function in the nvim-cmp, in mlua i made it works with the same signature, here i have a double free error. maybe i create Array incorrectly or i do something weird in the func?
Im trying to create something like that
arr = { { "label" = "Jan"} } where it is in index 0.

the same in mlua:
![image](https://user-images.githubusercontent.com/69643410/197339194-42ded6a4-04f4-4d72-8f77-e3f94082abe9.png)

![image](https://user-images.githubusercontent.com/69643410/197339281-2424352f-6b6c-4bfa-affd-00926ab83d93.png)

![image](https://user-images.githubusercontent.com/69643410/197339128-21e18bbb-9dd4-47d8-adb0-4ef4eabbb3fe.png)

**Comment by noib3** [comment]
Do you have a self-contained code snippet that reproduces the issue to figure out what's going wrong?

**Comment by yotam5** [comment]
> Do you have a self-contained code snippet that reproduces the issue to figure out what's going wrong?

if i get what you ask for here it is with some additional info if you need:

the code i wrote, both in lua and mlua:

lib.rs: https://pastebin.com/RYJ5L6Vi
Cargo.toml: https://pastebin.com/ZrTi1NMj

if you want to see the video i followed, ignore the new function: https://www.youtube.com/watch?v=om7r3PXT1Vg

if so just try to do so from the nvim-cmp i didnt change anything else

i posted the relevant part of the doc here, its very short, here: https://pastebin.com/jf9pnFmC

the doc, Ctrl + F the word "Develop": https://github.com/hrsh7th/nvim-cmp/blob/main/doc/cmp.txt

the lua code the worked, what i wrote without some optional section: https://pastebin.com/bfJt2R8s

**Comment by yotam5** [comment]
the Dictionary key is only a string, can it be anything that implement Hash or to_string?
from what i know know the only way to get lets say dict[i32] = "something" is Array, can this functionality be added to dict with what i said above or i miss something about the implementation that require only String.

ill delv more into the implementation later

**Comment by yotam5** [comment]
@noib3 
I did some debugging and the double free happens at functions.rs line 61 in the picture
![image](https://user-images.githubusercontent.com/69643410/197506197-a0c20af0-e956-4588-a9b4-2b4eb24e81fc.png)

i printed the args debug info before line 61 and it looks like this:

![image](https://user-images.githubusercontent.com/69643410/197507221-eb46f3cd-ee32-4b63-bac3-52eeecd6496b.png)

and its type is oxi object from the following lines:
![image](https://user-images.githubusercontent.com/69643410/197507340-4ecd0a0c-02bd-4408-84c8-302e01ecae52.png)

**Comment by noib3** [comment]
> if i get what you ask for here it is with some additional info if you need:

I was asking for something like [this](https://github.com/noib3/nvim-oxi/issues/70#issue-1410365776), i.e. a code snippet which is self-contained and only includes the code required to reproduce the error.

Trying to debug this by following screenshots or recreating plugins is a nightmare.

**Comment by noib3** [comment]
@yotam5 I was able to get it to crash with just:

```rust
use nvim_oxi::{self as oxi, Array, Dictionary, Function};

fn complete(func: Function<Array, ()>) -> oxi::Result<i32> {
    func.call(Array::from_iter(["hey"]))?;
    Ok(42)
}

#[oxi::module]
fn oxi() -> oxi::Result<Dictionary> {
    Ok(Dictionary::from_iter([(
        "complete",
        Function::from_fn(complete),
    )]))
}
```

and this `config.lua`:

```lua
vim.opt.rtp:append({ "<path_to_project>" })

local oxi = require("oxi")

local answer = oxi.complete(function(obj) print(obj); end)
print(answer)
```

Is this the problem you're referring to?

**Comment by noib3** [comment]
should be fixed by 29bdd3383eedc24bf86d733fa70dbfa21ac91d82
